### ISSUE TITLE
The threadCache is always missing the first couple threads in the Inbox

### ISSUE THREAD

**Issue by ek23** [KartikTalwar/gmail.js#589]
It looks like Google has recently made a change to how it retrieves threads when first loading Gmail.  The first i/fd XHR call asks for thread number six in the inbox.  The later calls ask for emails further down in the folder.  There is never an i/fd call asking for the first five threads.  Those first five threads are actually returned with the request to https://mail.google.com/mail/u/0/data.  

Has anyone else noticed this.  Seems like gmail.js would need to intercept the results of that /data call or just use the dom exclusively for those first five threads.

**Comment by josteink** [comment]
Hey there and thanks for the report!

Also: Sorry about the slow reply. Been really busy putting out fires elsewhere.

That said: I've always noticed that the first email (when loading straight into an email-view) is missing from the threadCache, but never considered it crucial enough to investigate further.

If now more emails are starting to miss from the threadCache, I guess that will lead to a perception about the new GmailJS Data-API not being sufficiently stable.

> use the dom exclusively for those first five threads.

DOM is sadly insufficient to get the data-granularity we want. If the DOM-data is sufficient, I would always advice to use the DOM-API directly since it's probably going to be more reliable, not to mention faster.

> Seems like gmail.js would need to intercept the results of that /data call

That's probably a better idea. I'll have to research how though: This call is not  a regular XHR call, but seems to be a regular web-request so the XHR interceptor wont work.

Finding out if this is possible, and if so exactly how, will probably take some research.

You already seems to have looked a bit into this. Would you care to dig a little more? :smile:

**Comment by josteink** [comment]
Also: Can confirm the most recent items in the Inbox is unavailable in the threadCache. 

This will definitely go noticed by many users in many extension, and should IMO really be fixed.

**Comment by josteink** [comment]
Doing some very basic testing, I just added the following patch to log all intercepted requests:

````diff
diff --git a/src/gmail.js b/src/gmail.js
index c9803f2..8d8a40d 100755
--- a/src/gmail.js
+++ b/src/gmail.js
@@ -1668,6 +1668,7 @@ var Gmail = function(localJQuery) {
                     var xhr = this;
                     this.onreadystatechange = function(progress) {
                         if (this.readyState === this.DONE) {
+                            console.log("INTERCEPTED REQUEST:" + this.xhrParams.url_raw);
                             if (progress.target.responseType === "" || progress.target.responseType === "text") {
                                 xhr.xhrResponse = api.tools.parse_response(progress.target.responseText);
                             } else {
````

And the bad news is that the `/data`-request doesn't show up, even with content-script set to load at `document_start`.

My guess is that we may have to re-architect GmailJS to run as part of the content-script, and extend the API-permissions required in `manifest.json` if we want to solve this:

````json
    "permissions": [
        "webRequest"
    ]
````

That of-course, will also require *re-writing every single extension which uses GmailJS*.

While I think that (in the long run) would be the cleanest, and most maintainable way to do things, I (thinking short term "lets just get this fixed") would rather not have to undertake that whole endeavour.

Any opinions one way or the other?

**Comment by josteink** [comment]
Not sure *when* this request gets loaded, or what pulls it in. Inspecting the DOM after main inbox is loaded yields nothing:

````javascript
var iframes = document.querySelectorAll("iframe")
var iframesrcs = Array.prototype.map.apply(iframes, [(i => i.src)]);
iframesrcs
(8) [
    "",
    "https://hangouts.google.com/webchat/u/0/load?clien…ps%3A%2F%2Fmail.google.com&ho...",
    "https://hangouts.google.com/webchat/u/0/frame2?v=1…...",
    "https://contacts.google.com/widget/hovercard/v/2?a…3A%...",
    "https://accounts.google.com/o/oauth2/postmessageRe…...",
    "https://hangouts.google.com/webchat/u/0/frame2?v=1...",
    "https://hangouts.google.com/webchat/u/0/frame?v=15…...",
    "https://hangouts.google.com/hangouts/_/hscv?pvt=AM…p..."
]
````

Nothing persisted for us to inspect.

I hoped we might manually be able to XHR that URL (https://mail.google.com/mail/u/0/data), but manually loading that plain URL without arguments in the brwoser, I get an empty data-set.

So what options does that leave us?

Any ideas?

**Comment by josteink** [comment]
Ok so I've looked into this once more, since 1. I have time and 2. it seemingly affects one of my extensions.

I can definitely confirm this behaviour, but I also see that trying to fetch data based on the legacy-ID still works for these emails.

That leads me implementing a workaround somewhat like this:

````javascript
const emailId = "msg-f:1642539442546720237";
const emailElem = document.querySelector(`[data-message-id='#${emailId}']`);
const emailLegacyId = emailElem.dataset.legacyMessageId
const emailLegacyData = gmail.get.email_data(emailLegacyId);
// TODO: extract data required by application from emailLegacyData 
````

This could/should probably be implemented inside Gmail.JS itself in the form of a async function like `api.new.get.email_data_with_fallback()` which also converts the data to the expected format.

Until then, this is the fallback I'm going to use to get something shipped :smile:

**Comment by Fabi1Sc** [comment]
Hi,

I think the problem come from the function api.tools.xhr_watcher.
Precisely where xhr response is parse and load in cache email_data

`// intercept email-data passively, instead of actively trying to fetch it later!
                            // (which we won't be able to do once 2019 hits anyway...)
                            if (api.check.is_new_data_layer()) {
                                if (api.tools.get_pathname_from_url(xhr.xhrParams.url_raw).endsWith("/i/fd")) {
                                    let parsed_emails = api.tools.parse_fd_request_payload(xhr.xhrResponse);
                                    api.tools.cache_email_data(parsed_emails);
                                    events.load_email_data = [parsed_emails];
                                }
                            }`

Only the pathname endwith with "/i/fd" is checked and when i debug xhr exchange i found the 5 first email in another pathname request. In my case it is :

          "method": "POST",
          "url": "https://mail.google.com/sync/u/0/i/s?hl=fr&c=9",

I think we have only to add this pathname and adapt the parser for this path if necessary.
Perhaps there are another pathname to check to have all messages...

What do you think ?

**Comment by josteink** [comment]
> Only the pathname endwith with "/i/fd" is checked and when i debug xhr exchange i found the 5 first email in another pathname request. In my case it is :

That's super-interesting.

If you can provide a PR with these changes, so that I can play around with it in the boilerplate project and verify myself, that's super-welcome.

We'll definitely approve any patches which helps solve this problem properly!

**Comment by Fabi1Sc** [comment]
Hi,

I will work on that on the next days or next week because this issue is a problem for me :)

By continuing my investigations, I clearly, il thing that bv request like "https://mail.google.com/sync/u/0/i/bv?hl=fr&c=0" contains more interesting informations than fd request like "https://mail.google.com/sync/u/0/i/fd?hl=fr&c=1" to analyze the inbox_content.

bv request contains full list of thread/email in the inbox_content and full list of label and it's possible to join label and thread/email with bv request.

The fd request is more interesting to have detail for each thread/email and to view thread or email on DOM.

**Comment by josteink** [comment]
Great. Thanks for looking into this!

When you do land some code based on your research, I would love a PR which fits nicely with what we have:

* keeps the new `gmail.new.get.email_data()` working exactly as before (from an API-consumer point of view), with whatever changes being done, being done transparantly in the back-end.
* whatever extra data you can get which you think is useful, feel free to extend the `gmail.new.*` namespace to include that functionality.

I guess for the first point, a review will mostly consider the code-form more than anything else. For the second, I guess we might go a little bit back and forth wrt what a good API should look like.

Therefore it might make sense to send a PR with the first change first and get that landed, before we start the discussions on the second bit. Sounds like a plan?

**Comment by Fabi1Sc** [comment]
OK I'm agree in principle.

I'll work on the first change because that's what i need.
I'll see where my investigation go depend on my time and the difficulty, by continuing my investigation, i see that request sequence is complex and depending on context like conversation mode on/off for example.

There are multiple xhr to synchronize thread and email data like :

- https://mail.google.com/sync/u/0/i/fd?hl=fr&c=1	
- https://mail.google.com/sync/u/0/i/bv?hl=fr&c=15
- https://mail.google.com/sync/u/0/i/s?hl=fr&c=17


I just find that the first request which return thread and email data is execute when loading Gmail and it's not XHR, it's a text/html content-type request on https://mail.google.com/mail/u/0/data?
The response is HTML with JSON embedded, that's why thread and email data can be display quickly without waiting XHR... 

I have the impression that email is append to cache only one time and cache never update... I'm not sure.

So its'complex....

**Comment by josteink** [comment]
Sounds complex indeed. Did you get anywhere with this? :)

**Comment by Fabi1Sc** [comment]
Yes i found and implemented a solution and it works.
Sorry for the late, i havent enough time now... confined at home with childs....
I think i can send a PR next week.

**Comment by josteink** [comment]
> Yes i found and implemented a solution and it works.

Great news!

> Sorry for the late, i havent enough time now... confined at home with childs....

No rush. I'm in the same situation and I understand perfectly well.

I was just curious if you got anywhere or not. Looking forward to see the results! :smiley:

**Comment by josteink** [comment]
This should now be fixed in the new and shiny `1.0.0` release!
